src/merge.rs

```diff
@@ -192,3 +194,295 @@ impl<'repo> Drop for AnnotatedCommit<'repo> {
+
+        self
+    }
+
+    /// File mode of the conflicted file, or `0` to not merge the mode.
+    pub fn mode(&mut self, mode: Option<FileMode>) -> &mut MergeFileInput {
+        self.mode = mode;
+
+        if let Some(mode) = self.mode {
+            self.raw.mode = mode as u32;
+        }
+
+        self
+    }
+
+    /// File content, text or binary
+    pub fn content(&mut self, content: Option<Vec<u8>>) -> &mut MergeFileInput {
+        self.content = content;
+
+        self.raw.size = self.content.as_ref().map(|c| c.len()).unwrap_or(0);
+        self.raw.ptr = self
+            .content
+            .as_ref()
+            .map(|c| c.as_ptr() as *const _)
+            .unwrap_or(ptr::null());
+
+        self
+    }
+
+    /// Get the raw struct in C
+    pub fn raw(&self) -> *const raw::git_merge_file_input {
+        &self.raw as *const _
+    }
+}
+
+impl MergeFileInput {
+    /// Create from Repository and IndexEntry
+    pub fn from(repo: &Repository, index_entry: &IndexEntry) -> MergeFileInput {
+        let blob = repo
+            .find_blob(index_entry.id.clone())
+            .expect("failed to find blob of index entry to make MergeFileInput");
+        let content = blob.content().to_vec();
+
+        let mut input = MergeFileInput::new();
+        input.content(Some(content));
+        input.path(index_entry.path.clone());
+        input.mode(Some(FileMode::from(index_entry.mode)));
+
+        input
+    }
+}
+
+impl std::fmt::Debug for MergeFileInput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let mut ds = f.debug_struct("MergeFileInput");
+        if let Some(path) = &self.path {
+            ds.field("path", path);
+        }
+        ds.field("mode", &self.mode);
+
+        if let Some(mode) = self.mode {
+            match mode {
+                FileMode::Unreadable => {}
+                FileMode::Tree => {}
+                FileMode::Blob => {
+                    let content = self
+                        .content
+                        .as_ref()
+                        .map(|s| String::from_utf8_lossy(&s).to_string())
+                        .unwrap_or("unknown content".to_string());
+                    ds.field("content", &content);
+                }
+                FileMode::BlobExecutable => {}
+                FileMode::Link => {}
+                FileMode::Commit => {}
+            }
+        }
+        ds.finish()
+    }
+}
+
+/// For git_merge_file_result
+pub struct MergeFileResult {
+    /// True if the output was automerged, false if the output contains
+    /// conflict markers.
+    pub automergeable: bool,
+
+    /// The path that the resultant merge file should use, or NULL if a
+    /// filename conflict would occur.
+    pub path: Option<String>,
+
+    /// The mode that the resultant merge file should use.
+    pub mode: FileMode,
+
+    /// The contents of the merge.
+    pub content: Option<Vec<u8>>,
```

Comment:
FWIW I think this would probably be best done by (privately) containing `raw::git_merge_file_result` and providing accessors for each field. That way Rust doesn't need to allocate anything unnecessarily.

Reply:
Great, improved.